include/IECore/MeshAlgo.h

```diff
@@ -47,11 +47,13 @@ namespace MeshAlgo
 {
 
 /// Calculate the surface tangent vectors of a mesh primitive.
-std::pair<PrimitiveVariable, PrimitiveVariable> calculateTangents( const MeshPrimitive *mesh,
-	const std::string &uvSet = "uv",
-	bool orthoTangents = true,
-	const std::string &position = "P"
-);
+std::pair<PrimitiveVariable, PrimitiveVariable> calculateTangents( const MeshPrimitive *mesh, const std::string &uvSet = "uv", bool orthoTangents = true, const std::string &position = "P" );
+
+/// Calculate the face area of a mesh primitive.
+PrimitiveVariable calculateFaceArea( const MeshPrimitive *mesh, const std::string &position = "P" );
+
+/// Calculate the face texture area of a mesh primitive based on the specified UV set.
+PrimitiveVariable calculateFaceTextureArea( const MeshPrimitive *mesh, const std::string &uvSet = "uv", const std::string &position = "P" );
```

Comment:
I did it this way because it seemed simpler to pass names, especially from python, and because `calculateTangents` was already doing it that way. For all these `calculate` algos, it seems to be implied that the variable is part of the primitive, otherwise the calculation wouldn't make sense (i.e why would you calculate texture area of a primitive using texture coordinates from a different primiitve?).

For `resamplePrimitiveVariable()` its explicitly about the variable rather than the primitive, and the resampling is likely in prep for adding to the primitive (rather than adding invalid data and then resampling in place), so that one makes sense to me.

To be honest, I found it a bit bizarre that `deleteFaces()` takes a `PrimitiveVariable` when all it needs is a VectorData. Someone who paints a deletion map is going to have to construct a `PrimitiveVariable` for no reason. Presumably that was to avoid templating or despatching on the client side?

@donboie would you like to weigh in for the new methods?

FYI, the FaceArea node in Gaffer will need to wrap both methods. They are independent calculations.

Reply:
> I did it this way because it seemed simpler to pass names, especially from python, and because calculateTangents was already doing it that way.

From Python it's super easy to pass primitive variables - `mesh["name"]`. And like I said, `calculateTangents()` was taking names for a very specific reason.

> For all these calculate algos, it seems to be implied that the variable is part of the primitive, otherwise the calculation wouldn't make sense (i.e why would you calculate texture area of a primitive using texture coordinates from a different primiitve?).

Agreed - that's unlikely. But that's not what I'm trying to get at - I'm talking about making the interface so that the client code is forced to consider what to do if variables are missing. Otherwise if the client code wants to be tolerant of errors, it's either a) doing it's own lookup first anyway, defeating the point of making the algo do it, or b) catching exceptions blindly and hoping they weren't about something else. Hence my question about what we think the behaviour of Gaffer nodes should be if requested variables are missing.

> To be honest, I found it a bit bizarre that deleteFaces() takes a PrimitiveVariable when all it needs is a VectorData.

True. It could take a DataPtr.

Reply:
Just discussing with @donboie over here. So far the best argument for me is by using string names with sensible defaults, all those calculate algo callsites are quite simple for the most common cases:

`result = MeshAlgo::functionName( mesh.get() )`

If we made them take PrimitiveVariables, then all the callsites need to be complex (e.g. no defaults).

`result = MeshAlgo::functionName( mesh.get(), uvPrimVar, pRefPrimVar, pPrimVar )`

The other point we made was maybe this doesn't need to be finalized in this massive PR. We could merge as is, and make another PR prior to an official Cortex 10 release that unifies them all.

Reply:
> So far the best argument for me is by using string names with sensible defaults, all those calculate algo callsites are quite simple for the most common cases.

Yeah, that's true. It's definitely nice when the caller knows that stuff _must_ be there, and an exception is the appropriate response to it not being there.

My main concern is Gaffer though, where the caller doesn't know what's coming from upstream, and we need to define what should happen when stuff is missing. If it should be a hard error in Gaffer, then the current interface of passing names and throwing is great. If it should be a silent no-op, the current interface is poor - the caller will need to duplicate all the checks that are currently inside the function, or unconditionally catch all exceptions, potentially silencing something unrelated but important. Neither of you have commented on that.

So, what do you think the appropriate behaviour is in Gaffer, as far as a user is concerned? I can see arguments both ways. Personally I want to see the error right away so I can correct it. But I imagine some users would prefer a more error-tolerant approach, where they can filter a node over a whole range of locations, and have it muddle through as best it can with what's there.

> The other point we made was maybe this doesn't need to be finalized in this massive PR. 

Agreed.

Reply:
> So, what do you think the appropriate behaviour is in Gaffer, as far as a user is concerned? I can see arguments both ways. Personally I want to see the error right away so I can correct it. But I imagine some users would prefer a more error-tolerant approach, where they can filter a node over a whole range of locations, and have it muddle through as best it can with what's there.

Coming from a Houdini background, I want both in Gaffer, by introducing a Warning state (to complement the Error state we have now). If my graph is:

`SceneReader -> FaceArea (filtered) -> OSLObject (filtered) -> Seeds (filtered)`

If I'm filtering a single object explicitly, I'd want the Error (red icon) on the FaceArea node when "uv" is missing. If I was filtering a bunch of objects, I'd probably want my stream to continue un-interrupted, but also non-silently (hence the per-object warning in yellow).

In either case, it seems the exception string thrown by the algo is likely to be the most useful bit of information. The node might still catch and re-raise to append some useful bits:

```
Warning : Invalid mesh at "/location/of/my/bad/mesh". MeshAlgo::distributePoints : MeshPrimitive has no suitable "uv" primitive variable.
```

We end up with some redundant bits maybe, but that would be an argument for an IECore::Message (to seperate out the prefix) rather than an exception I think.

Reply:
I see - that makes sense I guess. I guess you also want this warning status to propagate around to the downstream nodes and any parent nodes? That would require a bit of thought - we currently do it using exception propagation, but we won't be able to do that for warnings. I also don't know how we'll figure out when to make it an error and when to make it a warning - introspecting the filter doesn't seem right given that it's potentially very expensive to go looking for other matches in the scene.

Reply:
I don't know if they'd need to propogate as errors do... maybe they only show when you're explicitly inspecting that node in the graph?

Reply:
in houdini, when to error vs warn is a per-node plug. not sure if that's sensible for gaffer, but in the above network I might want to set my Seeds to error and my OSLObject to warn.